### PR TITLE
s390x: Add workaround for bsc#1083646

### PR DESCRIPTION
**Pull request by Soulofdestiny**
updated PR, updated Verification run

[This is an example how the workaround for bsc#1083646 would look like](http://localhost/tests/937)

@okurz

### CODE REVIEW

File: lib/utils.pm [hunk 1]
```diff
@@ -1156,11 +1154,32 @@ sub reconnect_s390 {
     }
     else {
         my $r = wait_serial($login_ready, 300);
-        if ($r =~ qr/Welcome to SUSE Linux Enterprise 15/) {
-            record_soft_failure('bsc#1040606');
-        }
-        elsif (is_sle) {
-            $r =~ qr/Welcome to SUSE Linux Enterprise Server/ || die "Correct welcome string not found";
+        if (is_sle) {
```
Comment (okurz) [inline lib/utils.pm]: The below all looks great but better not hide it in "is_sle". I guess the only SLE-specific part is the welcome message itself.

File: lib/utils.pm [hunk 2]
```diff
@@ -1156,11 +1154,32 @@ sub reconnect_s390 {
...
+            if ($r =~ qr/Welcome to SUSE Linux Enterprise Server/) {
+                return;
+            }
+            else {
+                opensusebasetest::record_info('Welcome message not found', 'Welcome string not found, try ssh to see what is going on...');
```
Comment (okurz) [inline lib/utils.pm]: opensusebasetest::record_info? The command record_info comes from the testapi itself. Where have you found this?

File: lib/utils.pm [hunk 3]
```diff
@@ -1156,11 +1154,32 @@ sub reconnect_s390 {
...
+                # Try to login via ssh
+                select_console 'root-console' || die "Correct welcome string not found, ssh login was not possible";
```
Comment (okurz) [inline lib/utils.pm]: Doesn't select_console die already anyway when it fails?

File: lib/utils.pm [hunk 4]
```diff
@@ -1156,11 +1154,32 @@ sub reconnect_s390 {
...
+                my $ret;
+                $ret = script_output 'systemctl status';
+                if ($ret =~ /0 queued/) {
+                    diag "No systemd jobs queued";
+                    return;
+                }
+                else {
```
Comment (okurz) [inline lib/utils.pm]: I don't see the need for the else when you already returned in before. I would write the above as

```perl
return diag 'No systemd jobs queued' if $ref =~ /0 queued/;
```

but maybe a record_info is even better? Or a "die" which will also put a text thumbnail on the details view? Or what should we try to continue in the job after we detected no jobs are queued? Like `die 'Welcome message not found but unknown error, no known mitigation'`?

File: lib/utils.pm (hunk 2 above)
Comment (Soulofdestiny) [reply lib/utils.pm]: really good question =)

File: lib/utils.pm (hunk 3 above)
Comment (Soulofdestiny) [reply lib/utils.pm]: true

File: lib/utils.pm [hunk 5]
```diff
@@ -1155,12 +1173,10 @@ sub reconnect_s390 {
...
     }
     else {
-        my $r = wait_serial($login_ready, 300);
-        if ($r =~ qr/Welcome to SUSE Linux Enterprise 15/) {
-            record_soft_failure('bsc#1040606');
-        }
-        elsif (is_sle) {
-            $r =~ qr/Welcome to SUSE Linux Enterprise Server/ || die "Correct welcome string not found";
+        my $r;
+        eval { $r = wait_serial($login_ready, $args{timeout}); };
```
Comment (okurz) [inline lib/utils.pm]: why do we need an eval? A wait_serial doesn't die, or does it?

File: lib/utils.pm [hunk 6]
```diff
@@ -1145,22 +1143,32 @@ sub reconnect_s390 {
...
-            record_soft_failure('bsc#1040606');
+        my $r;
+        $r = wait_serial 'Started OpenSSH Daemon', $args{timeout};
+        record_info 'no SSH Daemon', 'SSH Daemon was not started' && die "SSH Daemon was not started" if !$r;
+        wait_serial($login_ready, 30);
+        record_info('Welcome message not found', 'Welcome string not found, try ssh to see what is going on...');
+        # Try to login via ssh
+        select_console 'root-console';
+        my $ret;
+        $ret = script_output 'systemctl status';
```
Comment (rwx788) [inline lib/utils.pm]: Why not to initialize variable when declare?

File: lib/utils.pm [hunk 7]
```diff
@@ -1145,22 +1143,32 @@ sub reconnect_s390 {
     if (check_var('BACKEND', 's390x')) {
         my $r;
         eval { $r = console('x3270')->expect_3270(output_delim => $login_ready, timeout => $args{timeout}); };
-        if ($@) {
-            my $ret = $@;
-            _handle_login_not_found($ret);
-        }
+        _handle_login_not_found($@) if $@;
```
[the remark this answers is not in the document]
Comment (Soulofdestiny) [reply lib/utils.pm]: good point - I've rewritten this in on go when I tried the same approach for KVM but this didn't work at all. so I will either crosscheck this or revert - thanks

File: lib/utils.pm [hunk 8]
```diff
@@ -1145,22 +1143,32 @@ sub reconnect_s390 {
...
         select_console('iucvconn');
     }
     else {
-        my $r = wait_serial($login_ready, 300);
-        if ($r =~ qr/Welcome to SUSE Linux Enterprise 15/) {
-            record_soft_failure('bsc#1040606');
+        my $r;
+        $r = wait_serial 'Started OpenSSH Daemon', $args{timeout};
+        record_info 'no SSH Daemon', 'SSH Daemon was not started' && die "SSH Daemon was not started" if !$r;
+        wait_serial($login_ready, 30);
```
[the remark this answers is not in the document]
Comment (Soulofdestiny) [reply lib/utils.pm]: hmm it should continue, trying to ssh anyway and will fail on select_console?

[the remark this answers is not in the document]
Comment (Soulofdestiny) [reply lib/utils.pm]: hmm - it should still continue and sshing. Maybe another record_info would be more obvious here?

[the remark this answers is not in the document]
Comment (rwx788) [reply lib/utils.pm]: We have seen this behavior, it's same as script_run, shows red frame for the step and continues executing test module.

**Comment by Soulofdestiny** [comment]
updated.
[Verification run](http://opeth.suse.de/tests/939)

**Comment by Soulofdestiny** [comment]
```
[11:43] < mgriessmeie> | okurz: any objections to merge the workaround for the plymouth issue on reconnect_s390?
[12:53] <      okurz> | mgriessmeier: go ahead
```

**Comment by okurz** [comment]
Sorry, had to revert. Please re-revert. Then we should test again if we actually see the line "Started OpenSSH".

BTW, I have seen it in action but the PR description mentions a "localhost"-job, not that helpful ;)

**Comment by Soulofdestiny** [comment]
thanks - fixed the link
Will check why it doesn't work in production...

**Comment by Soulofdestiny** [comment]
apparently `Started OpenSSH Daemon` is not shown in openQA reliably...
I will come up with a better check and also try to implement the workaround into z/VM tests
-> tracked in [poo34609](https://progress.opensuse.org/issues/34609)

**Pull request merged**